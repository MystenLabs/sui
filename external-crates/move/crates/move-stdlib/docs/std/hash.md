
<a name="std_hash"></a>

# Module `std::hash`

Module which defines SHA hashes for byte vectors.

The functions in this module are natively declared both in the Move runtime
as in the Move prover's prelude.


-  [Function `sha2_256`](#std_hash_sha2_256)
-  [Function `sha3_256`](#std_hash_sha3_256)


<pre><code></code></pre>



<a name="std_hash_sha2_256"></a>

## Function `sha2_256`



<pre><code><b>public</b> <b>fun</b> <a href="../std/hash.md#std_hash_sha2_256">sha2_256</a>(data: <a href="../std/vector.md#std_vector">vector</a>&lt;<a href="../std/u8.md#std_u8">u8</a>&gt;): <a href="../std/vector.md#std_vector">vector</a>&lt;<a href="../std/u8.md#std_u8">u8</a>&gt;
</code></pre>



<details>
<summary>Implementation</summary>


<pre><code><b>native</b> <b>public</b> <b>fun</b> <a href="../std/hash.md#std_hash_sha2_256">sha2_256</a>(data: <a href="../std/vector.md#std_vector">vector</a>&lt;<a href="../std/u8.md#std_u8">u8</a>&gt;): <a href="../std/vector.md#std_vector">vector</a>&lt;<a href="../std/u8.md#std_u8">u8</a>&gt;;
</code></pre>



</details>

<a name="std_hash_sha3_256"></a>

## Function `sha3_256`



<pre><code><b>public</b> <b>fun</b> <a href="../std/hash.md#std_hash_sha3_256">sha3_256</a>(data: <a href="../std/vector.md#std_vector">vector</a>&lt;<a href="../std/u8.md#std_u8">u8</a>&gt;): <a href="../std/vector.md#std_vector">vector</a>&lt;<a href="../std/u8.md#std_u8">u8</a>&gt;
</code></pre>



<details>
<summary>Implementation</summary>


<pre><code><b>native</b> <b>public</b> <b>fun</b> <a href="../std/hash.md#std_hash_sha3_256">sha3_256</a>(data: <a href="../std/vector.md#std_vector">vector</a>&lt;<a href="../std/u8.md#std_u8">u8</a>&gt;): <a href="../std/vector.md#std_vector">vector</a>&lt;<a href="../std/u8.md#std_u8">u8</a>&gt;;
</code></pre>



</details>


[//]: # ("File containing references which can be used from documentation")
