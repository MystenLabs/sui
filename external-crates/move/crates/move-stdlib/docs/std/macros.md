
<a name="std_macros"></a>

# Module `std::macros`

This module holds shared implementation of macros used in <code>std</code>




<pre><code></code></pre>


[//]: # ("File containing references which can be used from documentation")
