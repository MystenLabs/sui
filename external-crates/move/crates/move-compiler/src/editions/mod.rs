// Copyright (c) The Move Contributors
// SPDX-License-Identifier: Apache-2.0

//! This module controls feature gating and breaking changes in new editions of the source language

use std::{
    collections::{BTreeMap, BTreeSet},
    fmt::Display,
    str::FromStr,
};

use crate::{
    diag,
    diagnostics::Diagnostic,
    shared::{string_utils::format_oxford_list, CompilationEnv},
};
use move_ir_types::location::*;
use move_symbol_pool::Symbol;
use once_cell::sync::Lazy;
use serde::{Deserialize, Serialize};

//**************************************************************************************************
// types
//**************************************************************************************************

#[derive(PartialEq, Eq, Clone, Copy, Debug, PartialOrd, Ord)]
pub struct Edition {
    pub edition: Symbol,
    pub release: Option<Symbol>,
}

#[derive(PartialEq, Eq, Clone, Copy, Debug, PartialOrd, Ord)]
pub enum FeatureGate {
    NestedUse,
    PublicPackage,
    PostFixAbilities,
    StructTypeVisibility,
    Enums,
    DotCall,
    PositionalFields,
    LetMut,
    Move2024Optimizations,
    Move2024Keywords,
    BlockLabels,
    Move2024Paths,
    MacroFuns,
    Move2024Migration,
    SyntaxMethods,
    AutoborrowEq,
    CleverAssertions,
    NoParensCast,
    TypeHoles,
    Lambda,
    ModuleLabel,
}

#[derive(PartialEq, Eq, Clone, Copy, Debug, PartialOrd, Ord, Default)]
pub enum Flavor {
    #[default]
    Core,
    Sui,
}

pub const UPGRADE_NOTE: &str =
    "You can update the edition in the 'Move.toml', or via command line flag if invoking the \
    compiler directly.";

//**************************************************************************************************
// Entry
//**************************************************************************************************

/// Returns true if the feature is present in the given edition.
/// Adds an error to the environment.
pub fn check_feature_or_error(
    env: &mut CompilationEnv,
    edition: Edition,
    feature: FeatureGate,
    loc: Loc,
) -> bool {
    if !edition.supports(feature) {
        env.add_diag(create_feature_error(edition, feature, loc));
        false
    } else {
        true
    }
}

pub fn feature_edition_error_msg(edition: Edition, feature: FeatureGate) -> Option<String> {
    let supports_feature = edition.supports(feature);
    if !supports_feature {
        let valid_editions = valid_editions_for_feature(feature);
        let message =
            if valid_editions.is_empty() && Edition::DEVELOPMENT.features().contains(&feature) {
                format!(
                    "{} under development and should not be used right now.",
                    feature.error_prefix()
                )
            } else {
                format!(
                    "{} not supported by current edition '{edition}', \
                    only {} support this feature",
                    feature.error_prefix(),
                    format_oxford_list!("and", "'{}'", valid_editions)
                )
            };
        Some(message)
    } else {
        None
    }
}

pub fn create_feature_error(edition: Edition, feature: FeatureGate, loc: Loc) -> Diagnostic {
    assert!(!edition.supports(feature));
    let Some(message) = feature_edition_error_msg(edition, feature) else {
        panic!("Previous assert should have failed");
    };
    let mut diag = diag!(Editions::FeatureTooNew, (loc, message));
    diag.add_note(UPGRADE_NOTE);
    diag
}

pub fn valid_editions_for_feature(feature: FeatureGate) -> Vec<Edition> {
    Edition::VALID
        .iter()
        .filter(|e| e.supports(feature))
        .copied()
        .collect()
}

//**************************************************************************************************
// impls
//**************************************************************************************************

static SUPPORTED_FEATURES: Lazy<BTreeMap<Edition, BTreeSet<FeatureGate>>> =
    Lazy::new(|| BTreeMap::from_iter(Edition::ALL.iter().map(|e| (*e, e.features()))));

const E2024_ALPHA_FEATURES: &[FeatureGate] = &[];

const E2024_BETA_FEATURES: &[FeatureGate] = &[
    FeatureGate::NestedUse,
    FeatureGate::PublicPackage,
    FeatureGate::PostFixAbilities,
    FeatureGate::StructTypeVisibility,
    FeatureGate::DotCall,
    FeatureGate::PositionalFields,
    FeatureGate::LetMut,
    FeatureGate::Move2024Keywords,
    FeatureGate::BlockLabels,
    FeatureGate::Move2024Paths,
    FeatureGate::Move2024Optimizations,
    FeatureGate::SyntaxMethods,
    FeatureGate::AutoborrowEq,
    FeatureGate::NoParensCast,
    FeatureGate::MacroFuns,
    FeatureGate::TypeHoles,
    FeatureGate::CleverAssertions,
    FeatureGate::Lambda,
    FeatureGate::ModuleLabel,
    FeatureGate::Enums,
];

const DEVELOPMENT_FEATURES: &[FeatureGate] = &[];

const E2024_MIGRATION_FEATURES: &[FeatureGate] = &[FeatureGate::Move2024Migration];

impl Edition {
    pub const LEGACY: Self = Self {
        edition: symbol!("legacy"),
        release: None,
    };
    pub const E2024_ALPHA: Self = Self {
        edition: symbol!("2024"),
        release: Some(symbol!("alpha")),
    };
    pub const E2024_BETA: Self = Self {
        edition: symbol!("2024"),
        release: Some(symbol!("beta")),
    };
    pub const E2024_MIGRATION: Self = Self {
        edition: symbol!("2024"),
        release: Some(symbol!("migration")),
    };
    pub const DEVELOPMENT: Self = Self {
        edition: symbol!("development"),
        release: None,
    };

    const SEP: &'static str = ".";

    pub const ALL: &'static [Self] = &[
        Self::LEGACY,
        Self::E2024_ALPHA,
        Self::E2024_BETA,
        Self::E2024_MIGRATION,
        Self::DEVELOPMENT,
    ];
    pub const VALID: &'static [Self] = &[Self::LEGACY, Self::E2024_ALPHA, Self::E2024_BETA];

    pub fn supports(&self, feature: FeatureGate) -> bool {
        SUPPORTED_FEATURES.get(self).unwrap().contains(&feature)
    }

    // Intended only for implementing the lazy static (supported feature map) above
    fn prev(&self) -> Option<Self> {
        match *self {
            Self::LEGACY => None,
            Self::E2024_ALPHA => Some(Self::E2024_BETA),
            Self::E2024_BETA => Some(Self::LEGACY),
            Self::E2024_MIGRATION => Some(Self::E2024_BETA),
            Self::DEVELOPMENT => Some(Self::E2024_ALPHA),
            _ => self.unknown_edition_panic(),
        }
    }

    // Inefficient and should be called only to implement the lazy static
    // (supported feature map) above
    fn features(&self) -> BTreeSet<FeatureGate> {
        match *self {
            Self::LEGACY => BTreeSet::new(),
            Self::E2024_ALPHA => {
                let mut features = self.prev().unwrap().features();
                features.extend(E2024_ALPHA_FEATURES);
                features
            }
            Self::E2024_BETA => {
                let mut features = self.prev().unwrap().features();
                features.extend(E2024_BETA_FEATURES);
                features
            }
            Self::E2024_MIGRATION => {
                let mut features = self.prev().unwrap().features();
                features.extend(E2024_MIGRATION_FEATURES);
                features
            }
            Self::DEVELOPMENT => {
                let mut features = self.prev().unwrap().features();
                features.extend(DEVELOPMENT_FEATURES);
                features
            }
            _ => self.unknown_edition_panic(),
        }
    }

    fn unknown_edition_panic(&self) -> ! {
        panic!("{}", self.unknown_edition_error())
    }

    pub fn unknown_edition_error(&self) -> anyhow::Error {
        anyhow::anyhow!(
            "Unsupported edition \"{self}\". Current supported editions include: {}",
            format_oxford_list!("and", "\"{}\"", Self::VALID)
        )
    }
}

impl Flavor {
    pub const CORE: &'static str = "core";
    pub const SUI: &'static str = "sui";
    pub const ALL: &'static [Self] = &[Self::Core, Self::Sui];
}

impl FeatureGate {
    fn error_prefix(&self) -> &'static str {
        match self {
            FeatureGate::NestedUse => "Nested 'use' forms are",
            FeatureGate::PublicPackage => "'public(package)' is",
            FeatureGate::PostFixAbilities => "Postfix abilities are",
            FeatureGate::StructTypeVisibility => "Struct visibility modifiers are",
            FeatureGate::Enums => "Enums are",
            FeatureGate::DotCall => "Method syntax is",
            FeatureGate::PositionalFields => "Positional fields are",
            FeatureGate::LetMut => "'mut' variable modifiers are",
            FeatureGate::Move2024Optimizations => "Move 2024 optimizations are",
            FeatureGate::Move2024Keywords => "Move 2024 keywords are",
            FeatureGate::BlockLabels => "Block labels are",
            FeatureGate::Move2024Paths => "Move 2024 paths are",
            FeatureGate::MacroFuns => "'macro' functions are",
            FeatureGate::Move2024Migration => "Move 2024 migration is",
            FeatureGate::SyntaxMethods => "'syntax' methods are",
            FeatureGate::AutoborrowEq => "Automatic borrowing is",
            FeatureGate::CleverAssertions => "Clever `assert!`, `abort`, and `#[error]` are",
            FeatureGate::NoParensCast => "'as' without parentheses is",
            FeatureGate::TypeHoles => "'_' placeholders for type inference are",
            FeatureGate::Lambda => "lambda expressions are",
            FeatureGate::ModuleLabel => "'module' label forms (ending with ';') are",
        }
    }
}

//**************************************************************************************************
// Parsing/Deserialize
//**************************************************************************************************

impl FromStr for Edition {
    type Err = anyhow::Error;

    // Required method
    fn from_str(s: &str) -> anyhow::Result<Self> {
        let (edition, release) = if let Some((edition, release)) = s.split_once(Edition::SEP) {
            (edition, Some(release))
        } else {
            (s, None)
        };
        let edition = Edition {
            edition: Symbol::from(edition),
            release: release.map(Symbol::from),
        };
        if !Self::VALID.iter().any(|e| e == &edition) && edition != Edition::DEVELOPMENT {
            return Err(edition.unknown_edition_error());
        }
        Ok(edition)
    }
}

impl FromStr for Flavor {
    type Err = anyhow::Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        Ok(match s {
            Self::CORE => Self::Core,
            Self::SUI => Self::Sui,
            _ => anyhow::bail!(
                "Unknown flavor \"{s}\". Expected one of: {}",
                Self::ALL
                    .iter()
                    .map(|e| format!("\"{}\"", e))
                    .collect::<Vec<_>>()
                    .join(", ")
            ),
        })
    }
}

impl<'de> Deserialize<'de> for Edition {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        Edition::from_str(&String::deserialize(deserializer)?)
            .map_err(|e| serde::de::Error::custom(format!("{e}")))
    }
}

impl<'de> Deserialize<'de> for Flavor {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        Flavor::from_str(&String::deserialize(deserializer)?)
            .map_err(|e| serde::de::Error::custom(format!("{e}")))
    }
}

//**************************************************************************************************
// Display/Serialize
//**************************************************************************************************

impl Display for Edition {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match &self.release {
            None => write!(f, "{}", self.edition),
            Some(release) => write!(f, "{}{}{}", self.edition, Self::SEP, release),
        }
    }
}

impl Display for Flavor {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Flavor::Core => write!(f, "{}", Self::CORE),
            Flavor::Sui => write!(f, "{}", Self::SUI),
        }
    }
}

impl Serialize for Edition {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        serializer.serialize_str(&format!("{}", self))
    }
}

impl Serialize for Flavor {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        serializer.serialize_str(&format!("{}", self))
    }
}

//**************************************************************************************************
// traits
//**************************************************************************************************

impl Default for Edition {
    fn default() -> Self {
        Edition::LEGACY
    }
}
