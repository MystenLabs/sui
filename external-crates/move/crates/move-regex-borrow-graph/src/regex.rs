// Copyright (c) The Move Contributors
// SPDX-License-Identifier: Apache-2.0

use std::fmt::{Debug, Display};

/// An edge in the graph contains a set of these simple regular expressions. For now, this is
/// just a (potentially empty) list of labels, and a flag indicating if the last label is a dot
/// star. This is overly constrained to the current limitations in Move. When we add support for
/// recursive types, we will need a more general regular expression. Particularly, we will need to
/// be able to have regular expressions after the dot star. And we will need to star regular
/// expressions other than dot.
#[derive(Clone, PartialEq, Eq, PartialOrd, Ord)]
pub(crate) struct Regex<Lbl> {
    labels: Vec<Lbl>,
    ends_in_dot_star: bool,
}

#[derive(Clone, PartialEq, Eq, PartialOrd, Ord)]
pub(crate) enum Extension<Lbl> {
    /// An empty regular expression
    Epsilon,
    /// A single label
    Label(Lbl),
    /// A dot star (matching 0 or more labels)
    DotStar,
}

//**************************************************************************************************
// impls
//**************************************************************************************************

impl<Lbl> Regex<Lbl> {
    /// An empty regular expression
    pub(crate) fn epsilon() -> Self {
        Self {
            labels: vec![],
            ends_in_dot_star: false,
        }
    }

    /// A single label
    pub(crate) fn label(lbl: Lbl) -> Self {
        Self {
            labels: vec![lbl],
            ends_in_dot_star: false,
        }
    }

    /// A dot star
    pub(crate) fn dot_star() -> Self {
        Self {
            labels: vec![],
            ends_in_dot_star: true,
        }
    }

    /// Returns true if the regex is empty (no labels and no dot star)
    pub(crate) fn is_epsilon(&self) -> bool {
        self.labels.is_empty() && !self.ends_in_dot_star
    }

    /// Used internally for metering of the graph within the bytecode verifier. We consider even
    /// an empty regex to be of size 1. Then we add one for each label and one for the dot star.
    /// In short, `abstract_size` is an abstraction over the Rust allocation size of the regex.
    pub(crate) fn abstract_size(&self) -> usize {
        1 + self.labels.len() + (self.ends_in_dot_star as usize)
    }

    /// Internal helper for the public facing API
    pub(crate) fn query_api_path(&self) -> (Vec<Lbl>, bool)
    where
        Lbl: Clone,
    {
        (self.labels.clone(), self.ends_in_dot_star)
    }

    /// This concatenates an extension onto the end of the regular expression.
    /// If the regular expression ends in dot star, then there is no current need to keep track of
    /// the labels after. The additional precision of having labels after the dot star is not
    /// needed currently in Move.
    pub(crate) fn extend(mut self, ext: &Extension<Lbl>) -> Self
    where
        Lbl: Clone,
    {
        match ext {
            _ if self.ends_in_dot_star => self,
            Extension::Epsilon => self,
            Extension::Label(lbl) => {
                self.labels.push(lbl.clone());
                self
            }
            Extension::DotStar => {
                self.ends_in_dot_star = true;
                self
            }
        }
    }

    /// If self = pq, then remove_prefix(p) returns Some(q) for all possible q.
    ///
    /// The input `p` is restricted to being an `Extension`. This is a practical limitation that
    /// makes writing the algorithm easier.
    ///
    /// Examples where there is a prefix, p:
    /// Remove epsilon
    /// "".remove_prefix("") = [""]
    /// "a".remove_prefix("") = ["a"]
    /// "ab".remove_prefix("") = ["ab"]
    /// ".*".remove_prefix("") = [".*"]
    /// "a.*".remove_prefix("") = ["a.*"]
    ///
    /// Remove label
    /// "ab".remove_prefix("a") = ["b"]
    /// "abc".remove_prefix("a") = ["bc"]
    /// ".*".remove_prefix("a") = [".*"]
    /// "a.*".remove_prefix("a") = [".*"]
    /// "ab.*".remove_prefix("a") = ["b.*"]
    ///
    /// Remove dot star
    /// "".remove_prefix(".*") = [""]
    /// "a".remove_prefix(".*") = ["a", ""]
    /// "ab".remove_prefix(".*") = ["ab", "b", ""]
    /// "abc".remove_prefix(".*") = ["abc", "bc", "c", ""]
    /// ".*".remove_prefix(".*") = [".*"]
    /// "a.*".remove_prefix(".*") = ["a.*", ".*", ""] ==> optimized to [".*"]
    /// "ab.*".remove_prefix(".*") = ["ab.*", "b.*", ".*", ""] ==> optimized to [".*"]
    ///
    /// Cases where there is no prefix:
    /// "".remove_prefix("a") = []
    /// "a".remove_prefix("b") = []
    /// "ab".remove_prefix("b") = []
    /// "abc".remove_prefix("b") = []
    /// "a.*".remove_prefix("b") = []
    pub(crate) fn remove_prefix(&self, p: &Extension<Lbl>) -> Vec<Regex<Lbl>>
    where
        Lbl: Clone,
        Lbl: Eq,
    {
        let mut self_walk = self.walk();
        match p {
            Extension::Epsilon => {
                let result = self_walk.remaining().into_iter().collect::<Vec<_>>();
                debug_assert!(!result.is_empty());
                result
            }
            Extension::Label(l_p) => {
                match self_walk.peek() {
                    WalkPeek::EmptySet => {
                        debug_assert!(false);
                        vec![]
                    }
                    WalkPeek::Epsilon => {
                        // cannot remove l1 from epsilon
                        vec![]
                    }
                    WalkPeek::Label(l_self) => {
                        if l_p != l_self {
                            // cannot remove l_p if it doesn't match l_self
                            vec![]
                        } else {
                            // we remove l_p and return the remaining
                            self_walk.next();
                            let result = self_walk.remaining().into_iter().collect::<Vec<_>>();
                            debug_assert!(!result.is_empty());
                            result
                        }
                    }
                    // we can remove any prefix and still have dot star
                    WalkPeek::DotStar => vec![Self::dot_star()],
                }
            }
            Extension::DotStar => {
                if self.ends_in_dot_star {
                    // This is an optimization for the case where we have a list of labels
                    // with a dot star. If this ends in dot star we will eventually add it to
                    // the result set, and `r | .*` is equivalent to `.*` so we don't need the
                    // partial paths
                    return vec![Self::dot_star()];
                }
                let mut result = vec![];
                while let Some(rem) = self_walk.remaining() {
                    result.push(rem);
                    self_walk.next();
                }
                debug_assert!(matches!(self_walk.peek(), WalkPeek::EmptySet));
                result
            }
        }
    }

    fn walk(&self) -> Walk<Lbl> {
        if self.is_epsilon() {
            Walk::Epsilon
        } else {
            Walk::Regex {
                regex: self,
                idx: 0,
            }
        }
    }
}

impl<Lbl> Extension<Lbl> {
    /// If self = pq, then remove_prefix(p) returns Some(q) for all possible q.
    ///
    /// This is similar to `Regex::remove_prefix`, but with the restrictions flipped.
    ///
    /// Cases where there is a prefix, p:
    /// Removing from epsilon
    /// "".remove_prefix("") = [""]
    /// "".remove_prefix(".*") = [""]
    ///
    /// Removing from label
    /// "a".remove_prefix("") = ["a"]
    /// "a".remove_prefix("a") = [""]
    /// "a".remove_prefix(".*") = ["a", ""]
    /// "a".remove_prefix("a.*") = [""]
    ///
    /// Removing from dot star
    /// ".*".remove_prefix("") = [".*"]
    /// ".*".remove_prefix("a") = [".*"]
    /// ".*".remove_prefix("ab") = [".*"]
    /// ".*".remove_prefix(".*") = [".*"]
    /// ".*".remove_prefix("a.*") = [".*"]
    ///
    /// Cases where there is no prefix:
    /// "".remove_prefix("a") = []
    /// "".remove_prefix("ab") = []
    /// "".remove_prefix("a.*") = []
    /// "a".remove_prefix("b") = []
    /// "a".remove_prefix("bc.*") = []
    /// "a".remove_prefix("ab") = []
    pub(crate) fn remove_prefix(&self, p: &Regex<Lbl>) -> Vec<Regex<Lbl>>
    where
        Lbl: Clone,
        Lbl: Eq,
    {
        let mut p_walk = p.walk();
        match p_walk.peek() {
            WalkPeek::EmptySet => {
                debug_assert!(false);
                vec![]
            }
            WalkPeek::Epsilon => {
                // p = epsilon so q = self
                vec![self.clone().into_regex()]
            }
            WalkPeek::Label(l_p) => match self {
                Extension::Epsilon => {
                    // cannot remove l1 from epsilon
                    vec![]
                }
                Extension::Label(l_self) => {
                    if l_p != l_self {
                        // cannot remove l_p if it doesn't match l_self
                        vec![]
                    } else {
                        p_walk.next();
                        match p_walk.peek() {
                            WalkPeek::EmptySet => {
                                debug_assert!(false);
                                vec![]
                            }
                            WalkPeek::Epsilon | WalkPeek::DotStar => {
                                // p = l_p and q = epsilon (or the case where dot star is epsilon)
                                vec![Regex::epsilon()]
                            }
                            WalkPeek::Label(_) => {
                                // there is another label that we cannot remove from self
                                vec![]
                            }
                        }
                    }
                }
                // we can remove any prefix and still have dot star
                Extension::DotStar => vec![Regex::dot_star()],
            },
            WalkPeek::DotStar => {
                match self {
                    Extension::DotStar => vec![Regex::dot_star()],
                    Extension::Epsilon => {
                        // Consider the case where p=.*=epsilon so p and q are epsilon
                        vec![Regex::epsilon()]
                    }
                    Extension::Label(_) => {
                        // Two possibilities
                        // p = epsilon and q = self
                        // p = self and q = epsilon
                        vec![self.clone().into_regex(), Regex::epsilon()]
                    }
                }
            }
        }
    }

    pub(crate) fn into_regex(self) -> Regex<Lbl> {
        match self {
            Extension::Epsilon => Regex::epsilon(),
            Extension::Label(lbl) => Regex::label(lbl),
            Extension::DotStar => Regex::dot_star(),
        }
    }
}

// Helper for remove_prefix
enum Walk<'a, Lbl> {
    EmptySet,
    Epsilon,
    Regex { regex: &'a Regex<Lbl>, idx: usize },
}

// Helper for remove_prefix
enum WalkPeek<'a, Lbl> {
    EmptySet,
    Epsilon,
    Label(&'a Lbl),
    DotStar,
}

impl<Lbl> Walk<'_, Lbl> {
    fn peek(&self) -> WalkPeek<'_, Lbl> {
        match self {
            Walk::EmptySet => WalkPeek::EmptySet,
            Walk::Epsilon => WalkPeek::Epsilon,
            Walk::Regex { regex, idx } => {
                if *idx < regex.labels.len() {
                    WalkPeek::Label(&regex.labels[*idx])
                } else if regex.ends_in_dot_star {
                    WalkPeek::DotStar
                } else {
                    debug_assert!(false);
                    WalkPeek::Epsilon
                }
            }
        }
    }

    fn next(&mut self) {
        match self {
            Walk::EmptySet => {
                debug_assert!(false);
            }
            Walk::Epsilon => {
                *self = Walk::EmptySet;
            }
            Walk::Regex { regex, idx } => {
                if *idx < regex.labels.len() {
                    *idx += 1;
                }
                if *idx >= regex.labels.len() && !regex.ends_in_dot_star {
                    *self = Walk::Epsilon;
                }
            }
        }
    }

    fn remaining(&self) -> Option<Regex<Lbl>>
    where
        Lbl: Clone,
    {
        match self {
            Walk::EmptySet => None,
            Walk::Epsilon => Some(Regex::epsilon()),
            Walk::Regex { regex, idx } => {
                let idx = *idx;
                let labels = if idx < regex.labels.len() {
                    regex.labels[idx..].to_vec()
                } else {
                    debug_assert!(regex.ends_in_dot_star);
                    vec![]
                };
                let ends_in_dot_star = regex.ends_in_dot_star;
                Some(Regex {
                    labels,
                    ends_in_dot_star,
                })
            }
        }
    }
}

//**************************************************************************************************
// traits
//**************************************************************************************************

/// Helper macro to help with Debug and Display. Cannot be a function since we cannot write
/// `Lbl: Display OR Debug`
macro_rules! fmt_regex {
    ($f:expr, $path:expr) => {{
        let f = $f;
        let p = $path;
        if p.is_epsilon() {
            return write!(f, "ε");
        }
        let mut exts = p.labels.iter().peekable();
        while let Some(ext) = exts.peek() {
            // display the element
            ext.fmt(f)?;
            // advance the iterator
            exts.next();
            // if there is a next element, add a separator
            if exts.peek().is_some() {
                write!(f, ".")?;
            }
        }
        if p.ends_in_dot_star {
            // we use _ instead of . to avoid confusion with field access/extension
            write!(f, "_*")?;
        }
        Ok(())
    }};
}

impl<Lbl> Display for Regex<Lbl>
where
    Lbl: Display,
{
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        fmt_regex!(f, self)
    }
}

impl<Lbl> Debug for Regex<Lbl>
where
    Lbl: Debug,
{
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        fmt_regex!(f, self)
    }
}
