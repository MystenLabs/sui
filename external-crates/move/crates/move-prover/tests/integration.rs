use codespan_reporting::term::termcolor::Buffer;
use glob;
use move_compiler::editions::Flavor;
use move_package::{BuildConfig as MoveBuildConfig, ModelConfig};
use move_prover::run_move_prover_with_model;
use regex::Regex;
use std::path::{Path, PathBuf};

/// Runs the prover on the given file path and returns the output as a string
fn run_prover(file_path: &PathBuf) -> String {
    // the file_dir path is `tests`, make it as a Path
    let file_dir = Path::new("tests");
    let sources_dir = file_dir.join("sources");
    // create the sources_dir if it doesn't exist
    if !sources_dir.clone().exists() {
        std::fs::create_dir_all(sources_dir.clone()).unwrap();
    }

    // Extract the relative path from tests/inputs/
    let relative_path = file_path
        .strip_prefix(Path::new("tests/inputs"))
        .unwrap_or_else(|_| Path::new(file_path.file_name().unwrap()));

    // Join it to the sources directory
    let new_file_path = sources_dir.join(relative_path);

    // Create parent directories if needed
    if let Some(parent_dir) = new_file_path.parent() {
        std::fs::create_dir_all(parent_dir).unwrap();
    }

    std::fs::rename(file_path, &new_file_path).unwrap();

    let new_file_path_clone = new_file_path.clone();

    // Setup cleanup that will execute even in case of panic or early return
    let result = std::panic::catch_unwind(|| {
        // Set up the build config
        let mut config = MoveBuildConfig::default();
        config.default_flavor = Some(Flavor::Sui);
        config.verify_mode = true;
        config.silence_warnings = false; // Disable warning suppression

        // Try to build the model
        let result = match config.move_model_for_package_legacy(
            file_dir,
            ModelConfig {
                all_files_as_targets: false,
                target_filter: None,
            },
        ) {
            Ok(model) => {
                // Create prover options
                let mut options = move_prover::cli::Options::default();
                options.backend.sequential_task = true;
                options.backend.use_array_theory = true;
                options.backend.vc_timeout = 3000;

                // Use a buffer to capture output instead of stderr
                let mut error_buffer = Buffer::no_color();

                // Run the prover with the buffer to capture all output
                match run_move_prover_with_model(&model, &mut error_buffer, options, None) {
                    Ok(_) => {
                        let error_output =
                            String::from_utf8_lossy(&error_buffer.into_inner()).to_string();
                        format!("Verification successful\n{}", error_output)
                    }
                    Err(err) => {
                        // Get the captured error output as string
                        let error_output =
                            String::from_utf8_lossy(&error_buffer.into_inner()).to_string();
                        format!("{}\n{}", err, error_output)
                    }
                }
            }
            Err(err) => {
                // For model-building errors, we need to reformat the error to match the expected format
                format!("We hit an error: \n{}", err)
            }
        };

        post_process_output(result)
    });

    // rename the file_path to the original name
    std::fs::rename(new_file_path_clone, file_path).unwrap();

    // Now handle the result of our operation
    match result {
        Ok(output) => output,
        Err(_) => "Verification failed: panic during verification".to_string(),
    }
}

fn post_process_output(output: String) -> String {
    // replace numbers such as 52571u64 with ELIDEDu64 to avoid snapshot diffs
    let output = output.replace("tests/sources/", "tests/inputs/");

    // Use regex to replace numbers with more than one digit followed by u64 with ELIDEDu64
    let re = Regex::new(r"\d{2,}u64").unwrap();
    re.replace_all(&output, "ELIDEDu64").to_string()
}

#[test]
fn run_move_tests() {
    for entry in glob::glob("tests/inputs/**/*.move").expect("Invalid glob pattern") {
        let move_path = entry.expect("Failed to read file path");
        let output = run_prover(&move_path);
        let filename = move_path.file_name().unwrap().to_string_lossy().to_string();

        let cp = move_path
            .parent()
            .unwrap()
            .components()
            .skip(2)
            .collect::<Vec<_>>();
        let cp_str = cp
            .iter()
            .map(|comp| comp.as_os_str().to_string_lossy().into_owned())
            .collect::<Vec<String>>();
        let snapshot_path = format!("snapshots/{}", cp_str.join("/"));

        insta::with_settings!({
            prepend_module_to_snapshot => false,
            snapshot_path => snapshot_path,
        }, {
            insta::assert_snapshot!(filename, output);
        });
    }
}
