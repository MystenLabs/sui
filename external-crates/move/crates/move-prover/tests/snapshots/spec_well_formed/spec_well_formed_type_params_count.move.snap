---
source: crates/move-prover/tests/integration.rs
expression: output
snapshot_kind: text
---
exiting with bytecode transformation errors
error: Spec function have differ type params count than underlying func
   ┌─ tests/inputs/spec_well_formed/spec_well_formed_type_params_count.move:11:1
   │  
11 │ ╭ public fun foo_spec<T, K>() {
12 │ │   foo<T>();
13 │ │   ensures(true);
14 │ │ }
   │ ╰─^
